crypto/fipsmodule/ec/ec_key.c

```diff
@@ -329,92 +329,81 @@ int EC_KEY_check_key(const EC_KEY *eckey) {
-      !EVP_DigestVerify(ctx, sig_der, sign_len, msg, msg_len)) {
-    goto err;
+
+  // Now that we have the coordinates we perform the range check.
+  if (BN_is_negative(x) || BN_cmp(x, &group->field) >= 0 ||
+      BN_is_negative(y) || BN_cmp(y, &group->field) >= 0) {
+      OPENSSL_PUT_ERROR(EC, EC_R_COORDINATES_OUT_OF_RANGE);
+    goto end;
   }
+
   ret = 1;
-err:
-  EVP_PKEY_free(evp_pkey);
-  OPENSSL_free(sig_der);
-  EVP_MD_CTX_free(ctx);
+end:
+  BN_free(x);
+  BN_free(y);
   return ret;
 }
 
+static int ec_key_ecdsa_pairwise_consistency_test(const EC_KEY *key) {
+  // Pairwise consistency test checks if a message signed by the private key
+  // can be correctly verified by the public key.
+  uint8_t data[16] = {0};
+  ECDSA_SIG *sig = ECDSA_do_sign(data, sizeof(data), key);
+  if (boringssl_fips_break_test("ECDSA_PWCT")) {
+    data[0] = ~data[0];
+  }
+  int ok = sig != NULL &&
+           ECDSA_do_verify(data, sizeof(data), sig, key);
+  ECDSA_SIG_free(sig);
+
+  return ok;
+}
+
 int EC_KEY_check_fips(const EC_KEY *key) {
-  // We have to avoid the underlying |EVP_DigestSign| and |EVP_DigestVerify|
-  // services in |EVP_EC_KEY_check_fips| updating the indicator state, so we
-  // lock the state here.
-  FIPS_service_indicator_lock_state();
   int ret = 0;
+  FIPS_service_indicator_lock_state();
 
-  if (EC_KEY_is_opaque(key)) {
+  if (EC_KEY_is_opaque(key) ||
+      !EC_KEY_check_key(key) ||
+      !ec_key_coordinates_range_check(key)) {
     // Opaque keys can't be checked.
     OPENSSL_PUT_ERROR(EC, EC_R_PUBLIC_KEY_VALIDATION_FAILED);
     goto end;
   }
 
-  if (!EC_KEY_check_key(key)) {
-    goto end;
-  }
-
-  // Check that the coordinates are within the range [0,p-1], when the (raw)
-  // point is affine; i.e. Z=1.
-  // This is the case when validating a received public key.
-  // Note: The check for x and y being negative seems superfluous since
-  // ec_felem_to_bignum() calls BN_bin2bn() which sets the `neg` flag to 0.
-  EC_POINT *pub_key = key->pub_key;
-  EC_GROUP *group = key->pub_key->group;
-  if(ec_felem_equal(group, &group->one, &pub_key->raw.Z)) {
-    BIGNUM *x = BN_new();
-    BIGNUM *y = BN_new();
-    int check_ret = 1;
-    if (group->meth->felem_to_bytes == NULL) {
-      OPENSSL_PUT_ERROR(EC, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
-      check_ret = 0;
-    } else if (!ec_felem_to_bignum(group, x, &pub_key->raw.X) ||
-               !ec_felem_to_bignum(group, y, &pub_key->raw.Y)) {
-      // Error already written to error queue by |bn_wexpand|.
-      check_ret = 0;
-    } else if (BN_is_negative(x) || BN_is_negative(y) ||
-               BN_cmp(x, &group->field) >= 0 ||
-               BN_cmp(y, &group->field) >= 0) {
-      OPENSSL_PUT_ERROR(EC, EC_R_COORDINATES_OUT_OF_RANGE);
-      check_ret = 0;
-    }
-    BN_free(x);
-    BN_free(y);
-    if (check_ret == 0) {
-      goto end;
-    }
-  }
-
   if (key->priv_key) {
-    if (!EVP_EC_KEY_check_fips((EC_KEY*)key)) {
```

Comment:
I should have probably noticed and asked this question much earlier; when `key->priv_key` is true, we do this pairwise consistency test (PCT) which uses signing and verifying. But in l.398, where `EC_KEY_check_key()` was called, another PCT was performed using multiplying the private key by the base point and comparing the public key. The latter agrees with [SP 800-56Ar3, Sec 5.6.2.1.4 Owner Assurance of Pair-wise Consistency](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf#%5B%7B%22num%22%3A108%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C96%2C289%2C0%5D). Do we still need the sign/verify PCT? Am I missing something?